WalletWasabi.Fluent.Generators/UIContextGenerator.cs

```diff
@@ -0,0 +1,332 @@
+
+			if (ctor.Parent is not ClassDeclarationSyntax cls)
+			{
+				continue;
+			}
+
+			if (!cls.IsRoutableViewModel(semanticModel))
+			{
+				continue;
+			}
+
+			var viewModelTypeInfo =
+				semanticModel.GetDeclaredSymbol(ctor.Parent as ClassDeclarationSyntax);
+
+			var className = cls.Identifier.ValueText;
+
+			var ctorNamespaces =
+					ctor.ParameterList.Parameters
+									  .Where(p => p.Type is not null)
+									  .Select(p => semanticModel.GetTypeInfo(p.Type!))
+									  .Where(t => t.Type is not null)
+									  .SelectMany(t => t.Type.GetNamespaces())
+									  .ToList();
+
+			var uiContextParam =
+				ctor.ParameterList
+					.Parameters
+					.FirstOrDefault(x => x.Type.IsUiContextType(semanticModel));
+
+			var methodParams = ctor.ParameterList;
+
+			if (uiContextParam != null)
+			{
+				methodParams = SyntaxFactory.ParameterList(methodParams.Parameters.Remove(uiContextParam));
+			}
+
+			var navigationMetadata =
+				viewModelTypeInfo.GetAttributes()
+						.FirstOrDefault(x => x.AttributeClass?.ToDisplayString() == NavigationMetaDataGenerator.NavigationMetaDataAttributeDisplayString);
+
+			var defaultNavigationTarget = "DialogScreen";
+
+			if (navigationMetadata != null)
+			{
+				var navigationArgument =
+					navigationMetadata.NamedArguments
+									  .FirstOrDefault(x => x.Key == "NavigationTarget");
+
+				if (navigationArgument.Value.Type is INamedTypeSymbol navigationTargetEnum)
+				{
+					var enumValue =
+						navigationTargetEnum.GetMembers()
+											.OfType<IFieldSymbol>()
+											.FirstOrDefault(x => x.ConstantValue?.Equals(navigationArgument.Value.Value) ?? false);
+
+					if (enumValue != null)
+					{
+						defaultNavigationTarget = enumValue.Name;
+					}
+				}
+			}
+
+			var additionalMethodParams =
+				$"NavigationTarget navigationTarget = NavigationTarget.{defaultNavigationTarget}, NavigationMode navigationMode = NavigationMode.Normal";
+
+			methodParams = methodParams.AddParameters(SyntaxFactory.ParseParameterList(additionalMethodParams).Parameters.ToArray());
+
+			var ctorArgs =
+				SyntaxFactory.ArgumentList(
+					SyntaxFactory.SeparatedList(
+					ctor.ParameterList
+						.Parameters
+						.Select(x => x.Type.IsUiContextType(semanticModel) ? "UiContext" : x.Identifier.ValueText) // replace uiContext argument for UiContext property reference
+						.Select(x => SyntaxFactory.ParseExpression(x))
+						.Select(SyntaxFactory.Argument),
+					ctor.ParameterList
+						.Parameters
+						.Skip(1)
+						.Select(x => SyntaxFactory.Token(SyntaxKind.CommaToken))));
+
+			namespaces.Add(viewModelTypeInfo.ContainingNamespace.ToDisplayString());
+			namespaces.AddRange(ctorNamespaces);
+
+			var methodName = className.Replace("ViewModel", "");
+
+			var methodString =
+$$"""
+    public void {{methodName}}{{methodParams}}
+	{
+	    UiContext.Navigate(navigationTarget).To(new {{className}}{{ctorArgs.ToFullString()}}, navigationMode);
+    }
+
+""";
```

Comment:
The same trick can be used here too, but the decision is up to you.

Reply:
@YohDeadfall see my comment above.

Reply:
Yeah, I feel your pain, but it isn't so bad as it could be (just 50 lines of code without fixing formatting; otherwise, even better).